casanovo/data/annotate_db.py

```diff
@@ -0,0 +1,131 @@
+
+        Parameters
+        ----------
+        tide_dir_path : str
+            Path to the directory containing the results of a successful tide search.
+        mgf_file : str
+            Path to the .mgf file containing MS/MS spectra.
+        output_file : str
+            Path to where the annotated .mgf will be written.
+
+    """
+    logger = logging.getLogger("casanovo")
+    # Get paths to tide search text files
+    tdf_path = os.path.join(tide_path, "tide-search.target.txt")
+    ddf_path = os.path.join(tide_path, "tide-search.decoy.txt")
+    try:
+        target_df = pd.read_csv(
+            tdf_path, sep="\t", usecols=["scan", "sequence", "target/decoy"]
+        )
+        decoy_df = pd.read_csv(
+            ddf_path, sep="\t", usecols=["scan", "sequence", "target/decoy"]
+        )
+    except FileNotFoundError as e:
+        logger.error(
+            "Could not find tide search results in the specified directory. "
+            "Please ensure that the directory contains the following files: "
+            "tide-search.target.txt and tide-search.decoy.txt"
+        )
+        raise e
+
+    logger.info("Successfully read tide search results from %s.", tide_path)
+
+    df = pd.concat([target_df, decoy_df])
+    scan_groups = df.groupby("scan")[["sequence", "target/decoy"]]
+
+    scan_map = {}
+
+    for scan, item in scan_groups:
+        td_group = item.groupby("target/decoy")["sequence"].apply(list)
+        if "target" in td_group.index:
+            target_candidate_list = list(
+                map(
+                    _normalize_mods,
+                    td_group["target"],
+                )
+            )
+        else:
+            target_candidate_list = []
+            logger.warn(f"No target peptides found for scan {scan}.")
+        if "decoy" in td_group.index:
+            decoy_candidate_list = list(
+                map(
+                    _normalize_mods,
+                    td_group["decoy"],
+                )
+            )
+            decoy_candidate_list = list(
+                map(lambda x: "decoy_" + str(x), decoy_candidate_list)
+            )
+        else:
+            decoy_candidate_list = []
+            logger.warn(f"No decoy peptides found for scan {scan}.")
+
+        scan_map[scan] = target_candidate_list + decoy_candidate_list
+
+    all_spec = []
+    for idx, spec_dict in enumerate(mgf.read(peak_path)):
+        try:
+            scan = int(spec_dict["params"]["scans"])
+        except KeyError as e:
+            logger.error(
+                "Could not find the scan number in the .mgf file. Please ensure that the .mgf file contains the scan number in the 'SCANS' field."
+            )
+            raise e
+        try:
+            spec_dict["params"]["seq"] = ",".join(list(scan_map[scan]))
+            all_spec.append(spec_dict)
+        except KeyError as e:
+            # No need to do anything if the scan is not found in the scan map
```

Comment:
What would cause this to occur? Tide not reporting any PSMs for a given peptide? We might want to log that.

Reply:
This would occur if a scan in the original `.mgf` was for some reason not considered by tide during its search. This is expected behavior and probably does not need to be logged. For example if we have a lot of low quality spectra and we set the `min_peaks` option to be higher, then many spectra would be skipped by tide and by extension Casanovo-DB.

Reply:
OK, then what we have now is fine. But when we add the tide functionality to Casanovo, I think we should be sure to report somewhere how many spectra were skipped/included like Tide does.